src/Native/Runtime/GCHelpers.cpp

```diff
@@ -220,3 +220,63 @@ COOP_PINVOKE_HELPER(void, RhGetMemoryInfo, (UInt32* highMemLoadThreshold, UInt64
                                                        lastRecordedMemLoad,
                                                        lastRecordedHeapSize, lastRecordedFragmentation);
 }
+
+EXTERN_C REDHAWK_API Int64 __cdecl RhGetTotalAllocatedBytes(Boolean precise)
+{
+    if (!precise)
+    {
+#if defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_)
+        uint64_t unused_bytes = Thread::dead_threads_non_alloc_bytes;
+#else
+        // As it could be noticed we read 64bit values that may be concurrently updated.
+        // Such reads are not guaranteed to be atomic on 32bit so extra care should be taken.
+        uint64_t unused_bytes = PalInterlockedCompareExchange64((Int64*)& Thread::dead_threads_non_alloc_bytes, 0, 0);
+#endif
+
+        uint64_t allocated_bytes = GCHeapUtilities::GetGCHeap()->GetTotalAllocatedBytes() - unused_bytes;
+
+        // highest reported allocated_bytes. We do not want to report a value less than that even if unused_bytes has increased.
+        static uint64_t high_watermark;
+
+        uint64_t current_high = high_watermark;
+        while (allocated_bytes > current_high)
+        {
+            uint64_t orig = PalInterlockedCompareExchange64((Int64*)& high_watermark, allocated_bytes, current_high);
+            if (orig == current_high)
+                return allocated_bytes;
+
+            current_high = orig;
+        }
+
+        return current_high;
+    }
+
+    Thread* pCurThread = ThreadStore::GetCurrentThread();
+    ASSERT(!pCurThread->IsCurrentThreadInCooperativeMode());
+
+    pCurThread->SetupHackPInvokeTunnel();
+    pCurThread->DisablePreemptiveMode();
```

Comment:
Not sure what to make about these two. They're used in the other `EXTERN_C` ones, so it appears to go managed coop -> unmanaged preemptive -> unmanaged coop, but why?